rotkehlchen/chain/ethereum/decoding/kyber/decoder.py

```diff
@@ -0,0 +1,95 @@
+        destination_token_address = hex_or_bytes_to_address(tx_log.data[32:64])
+
+        source_token = ethaddress_to_asset(source_token_address)
+        if source_token is None:
+            return None, None
+        destination_token = ethaddress_to_asset(destination_token_address)
+        if destination_token is None:
+            return None, None
+
+        if upgraded:
+            spent_amount_raw = hex_or_bytes_to_int(tx_log.data[96:128])
+            return_amount_raw = hex_or_bytes_to_int(tx_log.data[128:160])
+        else:
+            spent_amount_raw = hex_or_bytes_to_int(tx_log.data[64:96])
+            return_amount_raw = hex_or_bytes_to_int(tx_log.data[96:128])
+
+        spent_amount = asset_normalized_value(amount=spent_amount_raw, asset=source_token)
+        return_amount = asset_normalized_value(amount=return_amount_raw, asset=destination_token)
+
+        for event in decoded_events:
+            if event.event_type == HistoryEventType.SPEND and event.location_label == sender and event.asset == source_token and event.balance.amount == spent_amount:  # noqa: E501
+                event.event_type = HistoryEventType.TRADE
+                event.event_subtype = HistoryEventSubType.SPEND
+                event.counterparty = 'kyber legacy'
+                event.notes = f'Swap {event.balance.amount} {event.asset.symbol} in kyber'
+            elif event.event_type == HistoryEventType.RECEIVE and event.location_label == sender and event.balance.amount == return_amount and destination_token == event.asset:  # noqa: E501
+                event.event_type = HistoryEventType.TRADE
+                event.event_subtype = HistoryEventSubType.RECEIVE
+                event.counterparty = 'kyber legacy'
+                event.notes = f'Receive {event.balance.amount} {event.asset.symbol} from kyber swap'  # noqa: E501
+
+        return None, None
+
+    def decode_action(
+            self,
+            tx_log: EthereumTxReceiptLog,
+            transaction: EthereumTransaction,  # pylint: disable=unused-argument
+            decoded_events: List[HistoryBaseEntry],
+            all_logs: List[EthereumTxReceiptLog],  # pylint: disable=unused-argument
+            action_items: Optional[List[ActionItem]],  # pylint: disable=unused-argument
+    ) -> Tuple[Optional[HistoryBaseEntry], Optional[ActionItem]]:
+        if tx_log.topics[0] == KYBER_TRADE_LEGACY:
+            return self._decode_legacy_trade(tx_log=tx_log, decoded_events=decoded_events, upgraded=False)  # noqa: E501
+        if tx_log.topics[0] == KYBER_TRADE_LEGACY_UPGRADED:
+            return self._decode_legacy_trade(tx_log=tx_log, decoded_events=decoded_events, upgraded=True)  # noqa: E501
+        return None, None
+
+    def addresses_to_decoders(self) -> Dict[ChecksumEthAddress, Tuple[Any, ...]]:
+        return {
+            string_to_ethereum_address('0x9ae49C0d7F8F9EF4B864e004FE86Ac8294E20950'): (self.decode_action,),  # noqa: E501
```

Comment:
Hmmm this is not the correct approach.

The address in question is what should map to the decoder only for that address.

From what I understand here you have 2 different contracts.

1. First name them. Add them as named constants in the file
2. Each address should map to a different decoding functon.